### PR TITLE
K8SPG-38 Upgrade cluster test

### PR DESCRIPTION
**Pull request by cap1984**

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| demand-backup | passed |
| init-deploy | passed |
| data-migration-gcs | passed |
| scaling | passed |
| recreate | passed |
| affinity | passed |
| monitoring | passed |
| self-healing | passed |
| operator-self-healing | passed |
| scheduled-backup | failed |
| clone-cluster | passed |
| tls-check | passed |

commit: https://github.com/percona/percona-postgresql-operator/pull/127/commits/b26c1b2a931e5540c29cc2820c8081f524cbe0b3
image: `perconalab/percona-postgresql-operator:PR-127-b26c1b2a-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-127-b26c1b2a-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-127-b26c1b2a-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-127-b26c1b2a-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-127-b26c1b2a-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-127-b26c1b2a-pgo-deployer`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| demand-backup | passed |
| data-migration-gcs | passed |
| scaling | passed |
| recreate | passed |
| affinity | passed |
| monitoring | passed |
| self-healing | passed |
| scheduled-backup | passed |
| operator-self-healing | passed |
| clone-cluster | passed |
| tls-check | passed |

commit: https://github.com/percona/percona-postgresql-operator/pull/127/commits/e9b4b70bcce63fd7076c2305229a330b86913f47
image: `perconalab/percona-postgresql-operator:PR-127-e9b4b70b-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-127-e9b4b70b-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-127-e9b4b70b-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-127-e9b4b70b-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-127-e9b4b70b-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-127-e9b4b70b-pgo-deployer`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| demand-backup | passed |
| init-deploy | passed |
| data-migration-gcs | passed |
| scaling | passed |
| recreate | passed |
| affinity | passed |
| monitoring | passed |
| self-healing | passed |
| scheduled-backup | passed |
| operator-self-healing | passed |
| clone-cluster | passed |
| tls-check | passed |

commit: https://github.com/percona/percona-postgresql-operator/pull/127/commits/f34ee614e1f4a371390b5877696c247bbc6323c5
image: `perconalab/percona-postgresql-operator:PR-127-f34ee614-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-127-f34ee614-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-127-f34ee614-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-127-f34ee614-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-127-f34ee614-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-127-f34ee614-pgo-deployer`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| demand-backup | passed |
| data-migration-gcs | passed |
| scaling | passed |
| recreate | passed |
| scheduled-backup | passed |
| affinity | passed |
| monitoring | passed |
| self-healing | passed |
| operator-self-healing | passed |
| clone-cluster | passed |
| tls-check | passed |

commit: https://github.com/percona/percona-postgresql-operator/pull/127/commits/de04b9fc61f05986ed8f008473f653b9451e2985
image: `perconalab/percona-postgresql-operator:PR-127-de04b9fc-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-127-de04b9fc-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-127-de04b9fc-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-127-de04b9fc-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-127-de04b9fc-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-127-de04b9fc-pgo-deployer`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| demand-backup | passed |
| scaling | passed |
| scheduled-backup | passed |
| recreate | passed |
| affinity | passed |
| monitoring | passed |
| self-healing | passed |
| operator-self-healing | passed |
| clone-cluster | passed |
| tls-check | passed |
| data-migration-gcs | passed |

commit: https://github.com/percona/percona-postgresql-operator/pull/127/commits/de04b9fc61f05986ed8f008473f653b9451e2985
image: `perconalab/percona-postgresql-operator:PR-127-de04b9fc-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-127-de04b9fc-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-127-de04b9fc-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-127-de04b9fc-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-127-de04b9fc-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-127-de04b9fc-pgo-deployer`

File: e2e-tests/upgrade/run [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+. "${test_dir}"/../functions
+
+IMAGE_OPERATOR="percona/percona-postgresql-operator:0.2.0-postgres-operator"
+IMAGE_DEPLOYER="percona/percona-postgresql-operator:0.2.0-pgo-deployer"
+
+IMAGE_PGBOUNCER="percona/percona-postgresql-operator:0.2.0-ppg13-pgbouncer"
+IMAGE_PG_HA="percona/percona-postgresql-operator:0.2.0-ppg13-postgres-ha"
+IMAGE_BACKREST="percona/percona-postgresql-operator:0.2.0-ppg13-pgbackrest"
+IMAGE_BACKREST_REPO="percona/percona-postgresql-operator:0.2.0-ppg13-pgbackrest-repo"
+IMAGE_PGBADGER="percona/percona-postgresql-operator:0.2.0-ppg13-pgbadger"
```
Comment (hors) [inline e2e-tests/upgrade/run:16]: Should we have possibility of redefining all these vars?

File: e2e-tests/upgrade/run [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+compare_psql() {
```
Comment (hors) [inline e2e-tests/upgrade/run:18]: Can we move it to 'functions' file?

File: e2e-tests/upgrade/run [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+		>$tmp_dir/${command_id}.sql
+	if [ ! -s "$tmp_dir/${command_id}.sql" ]; then
+		sleep 20
+		run_psql_ssl "$command" "$uri" "postgres" "" "$sslmode" \
+			>$tmp_dir/${command_id}.sql
+	fi
+	diff -u $expected_result $tmp_dir/${command_id}.sql
+}
+
+run_psql_ssl() {
```
Comment (hors) [inline e2e-tests/upgrade/run:36]: Can we move it to 'functions' file?

File: e2e-tests/upgrade/run [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+			{"op":"replace","path":"/spec/pgBouncer/image","value":'${IMAGE_PGBOUNCER}'},
+			{"op":"replace","path":"/spec/pgPrimary/image","value":'${IMAGE_PG_HA}'},
+			{"op":"replace","path":"/spec/userLabels/pgo-version","value":'${GIT_BRANCH}'},
+			{"op":"replace","path":"/metadata/labels/pgo-version","value":'${GIT_BRANCH}'},
+			{"op":"replace","path":"/spec/pause","value":false}
+		]'
+
+	wait_cluster_consistency "${cluster}"
+	wait_deployment "${cluster}"
+	# wait for some significant time period. Pgbouncer pod may not be started otherwise
```
Comment (hors) [inline e2e-tests/upgrade/run:101]: As I understand we need to wait till primary pod is ready, am I right ?

File: e2e-tests/conf/ca.crt [hunk 5]
```diff
@@ -0,0 +1,10 @@
+-----BEGIN CERTIFICATE-----
```
Comment (hors) [inline e2e-tests/conf/ca.crt:1]: Can we use it from TLS test? Why do we need to have almost the same files in different dirs?

**Review by hors — CHANGES_REQUESTED** [review]

File: e2e-tests/conf/ca.crt (hunk 5 above)
Comment (cap1984) [reply e2e-tests/conf/ca.crt]: done

File: e2e-tests/upgrade/run (hunk 3 above)
Comment (cap1984) [reply e2e-tests/upgrade/run]: done

File: e2e-tests/upgrade/run (hunk 2 above)
Comment (cap1984) [reply e2e-tests/upgrade/run]: done

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| demand-backup | passed |
| data-migration-gcs | passed |
| scaling | passed |
| recreate | passed |
| affinity | passed |
| monitoring | passed |
| self-healing | passed |
| scheduled-backup | failed |
| operator-self-healing | passed |
| clone-cluster | passed |
| tls-check | passed |
| upgrade | failed |

commit: https://github.com/percona/percona-postgresql-operator/pull/127/commits/0d5bb0e6b2a69f1354bda77f210257e81d28d0c7
image: `perconalab/percona-postgresql-operator:PR-127-0d5bb0e6-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-127-0d5bb0e6-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-127-0d5bb0e6-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-127-0d5bb0e6-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-127-0d5bb0e6-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-127-0d5bb0e6-pgo-deployer`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| demand-backup | passed |
| scaling | passed |
| scheduled-backup | passed |
| recreate | passed |
| affinity | passed |
| monitoring | passed |
| self-healing | passed |
| operator-self-healing | passed |
| clone-cluster | passed |
| tls-check | passed |
| upgrade | passed |
| data-migration-gcs | passed |

commit: https://github.com/percona/percona-postgresql-operator/pull/127/commits/0d5bb0e6b2a69f1354bda77f210257e81d28d0c7
image: `perconalab/percona-postgresql-operator:PR-127-0d5bb0e6-pgo-apiserver`

image: `perconalab/percona-postgresql-operator:PR-127-0d5bb0e6-pgo-event`

image: `perconalab/percona-postgresql-operator:PR-127-0d5bb0e6-pgo-rmdata`

image: `perconalab/percona-postgresql-operator:PR-127-0d5bb0e6-pgo-scheduler`

image: `perconalab/percona-postgresql-operator:PR-127-0d5bb0e6-postgres-operator`

image: `perconalab/percona-postgresql-operator:PR-127-0d5bb0e6-pgo-deployer`

File: e2e-tests/upgrade/run (hunk 1 above)
Comment (hors) [reply e2e-tests/upgrade/run:16]: yes

File: e2e-tests/upgrade/run (hunk 4 above)
Comment (hors) [reply e2e-tests/upgrade/run:71]: strange way to get it, just wait 90s

**Review by hors — APPROVED** [review]

**Pull request merged**